Ensure `Matrix` * `(Sym)Tridiagonal` is dense

mcognetta opened the pull request:
This PR ensures that `Matrix` * `(Sym)Tridiagonal` outputs a dense matrix. Before, it was outputting a `SparseMatrixCSC`, despite the resulting matrix being dense.

This PR currently only implements this optimization for the `Matrix` type, so as to avoid method ambiguities, but it is possible that there are other type combinations that suffer from the same problem.

------------------------------------------------------------------------------------
A basic benchmark, showing a ~150x speedup:

Before:

```julia
julia> t = Tridiagonal(rand(1000, 1000)); s = SymTridiagonal(rand(1000), rand(999)); y = rand(1000, 1000);

julia> @btime $y * $t;
  256.924 ms (41 allocations: 18.01 MiB)

julia> @btime $y * $s;
  260.044 ms (41 allocations: 18.01 MiB)
```

This PR:

```julia
julia> t = Tridiagonal(rand(1000, 1000)); s = SymTridiagonal(rand(1000), rand(999)); y = rand(1000, 1000);

julia> @btime $y * $t;
  1.477 ms (2 allocations: 7.63 MiB)

julia> @btime $y * $s;
  1.469 ms (2 allocations: 7.63 MiB)
```

vchuravy committed:
Add Type to ByVal attribute

vchuravy committed:
[LLVM/Win32] Force stack alignment on module

vchuravy committed:
StackProtector is now a module flag

vchuravy committed:
Cleanup MachineObjectFileInfo handling in disassembly

vchuravy committed:
Disable NewPM for llvmpasses test

vtjnash committed:
[LateLowerGCFrame] handle undef values from a shuffle vector mask

The same as how we handle them if they were undef in the shuffle vector operand.

vtjnash committed:
fix tbaa_make_child use errors in our llvm passes

now caught by the verifier

dkarrasch committed:
Make LinearAlgebra.jl independent of SparseArrays.jl

dkarrasch committed:
add `sparse_*cat`

Commit:
[JuliaJITEventListener] Use llvm::StringMap

Commit:
[EE] CompilerUsed instead of Used

JeffBezanson committed:
make sure compilation signature is inferred when it differs from call site sig

JeffBezanson committed:
make sure precompile statements are printed even for const_return functions

mcognetta committed:
ensure dense * sym/tridiagonal is dense

mcognetta committed:
ensure dense * sym/tridiagonal is dense

mcognetta commented:
`buildbot/tester_freebsd64` check failure seems unrelated.

dkarrasch commented:
Could it be that this is already handled by #43127?

vchuravy committed:
SROA: Ensure that unused structs are preserved in ccall (#43408)

Co-authored-by: Shuhei Kadowaki <40514306+aviatesk@users.noreply.github.com>

dkarrasch commented:
> Could it be that this is already handled by #43127?

Yes, it is, I checked. Good to know that there are also performance benefits from splitting LinearAlgebra and SparseArrays. Maybe we keep this PR for the added tests until #43127 is merged?

dkarrasch commented:
BTW, the same change should be made for `Bidiagonal`, and also that is already "included" in #43127.

mcognetta commented:
Ah, thanks for the info. I didn't realize this was in the scope of that PR. I'll keep it around until #43127 is merged.

vchuravy committed:
Merge pull request #43417 from JuliaLang/vc/backport_llvm_13

[LLVM] Support LLVM 13.0.0

mcabbott committed:
Mention `norm` in docstring of `hypot` (#42972)

* mention norm in doc of hypot

Co-authored-by: Fredrik Ekre <ekrefredrik@gmail.com>
Co-authored-by: Viral B. Shah <ViralBShah@users.noreply.github.com>
Co-authored-by: Jameson Nash <vtjnash@gmail.com>

fp4code committed:
Fix typo: \nor is \barvee instead of \veebar (#43421)

dkarrasch committed:
Replace Val-types in cholesky[!] (#41640)

fonsp committed:
Manual: Move introduction about command line arguments from "Getting Started" to "Command-line options" (#42878)

DilumAluthgeBot committed:
🤖 Bump the Downloads stdlib from a55825b to c445f45 (#43430)

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

kshyatt committed:
Add test for IdSet pop with default (#43424)

aviatesk committed:
optimizer: use a more functional data flow for using constant calling convention (#43377)

Stop mutating `opt.const_api` field to indicate the eligibility of
constant calling convention.
Also this commit setups a specific data type `ConstAPI` for representing
constant values, which is just separated from the inference lattice type `Const`,
for more clarity of data flow.

Commit:
use IEEEFloat to clean up floats.jl (#43433)

stevengj commented on stdlib/LinearAlgebra/src/special.jl:
```diff
@@ -294,6 +294,12 @@ rmul!(A::AbstractTriangular, adjB::Adjoint{<:Any,<:Union{QRCompactWYQ,QRPackedQ}
 *(adjA::Adjoint{<:Any,<:Union{QRCompactWYQ, QRPackedQ}}, B::BiTriSym) =
     lmul!(adjA, copyto!(Array{promote_type(eltype(adjA), eltype(B))}(undef, size(B)...), B))
 
+function (*)(A::Matrix, B::Union{SymTridiagonal, Tridiagonal})
```
```suggestion
function (*)(A::StridedMatrix, B::Union{SymTridiagonal, Tridiagonal})
```
?

stevengj commented:
Is there a similar method for `tridiagonal * matrix` that also needs to be fixed?

mcognetta commented:
@stevengj Not for Tridiagonal * Dense, but perhaps for others. I think we are getting pretty close to catching all the defined structured matrix multiplication optimizations (though it always seems like one slips through the cracks).

chriselrod committed:
Add support for per-module max_methods. (#43370)

Co-authored-by: Tim Holy <tim.holy@gmail.com>
Co-authored-by: Shuhei Kadowaki <aviatesk@gmail.com>

aviatesk committed:
inference: some optimization improvements (#43447)

- cut off allocation in `tmerge`
- improve type instabilities

stevengj committed:
support Unicode 14.0.0 (via utf8proc 2.7.0) (#43443)

vtjnash committed:
Allow const declarations on mutable fields (#43305)

Mark some builtin types also, although Serialization relies upon being
able to mutilate the Method objects, so we do not yet mark those.

Replaces #11430

Co-authored-by: Matt Bauman <mbauman@gmail.com>

aviatesk committed:
Merge pull request #43415 from JuliaLang/jb/precompile2

some precompilation improvements

vtjnash committed:
[Serialization] support AbstractLock and GenericCondition (#43325)

Locks should be unlocked when serialized, and GenericCondition should
drop their waiters.

Otherwise, we may try to serialize a running Task (the user should
normally be holding the lock around the data they are intending to
serialize), which will fail and error.

krynju committed:
remove the copy of the future in serialize

gdsimoes committed:
Add percentage sign notation (#43460)

The rem function should have the percentage sign notation beside it in the division functions table.

vchuravy committed:
Merge pull request #43465 from krynju/kr/remove-future-copy-on-serialize

[Distributed] Remove (now) unnecessary copy of the Future in its serializer

ararslan committed:
Fix a case of `-Wmisleading-indentation` when building with Clang (#43472)

```
In file included from /usr/home/alex/julia/src/codegen.cpp:2005:
In file included from ./intrinsics.cpp:7:
In file included from ./ccall.cpp:322:
./abi_x86.cpp:78:9: warning: misleading indentation; statement is not part of the previous 'if' [-Wmisleading-indentation]
        ab.addByValAttr(Ty);
        ^
./abi_x86.cpp:73:5: note: previous statement is here
    if (is_complex64(dt) || is_complex128(dt) || (jl_is_primitivetype(dt) && size <= 8))
    ^
1 warning generated.
```

pchintalapudi committed:
Tweak bounds check elimination (#42692)

* Add GVN and move IRCE to allow more functions to vectorize
* Canonicalize arraysize(Vector, 1) to arraylen(Vector)

Co-authored-by: Prem Chintalapudi <premc@csail.mit.edu>
Co-authored-by: Valentin Churavy <vchuravy@users.noreply.github.com>

Commit:
add const keyword to broadcasted AndAnd (#43483)

Co-authored-by: Shuhei Kadowaki <40514306+aviatesk@users.noreply.github.com>

aviatesk committed:
inlining: add missing late special handling for `UnionAll` method call (#43479)

Looking at the result of <https://github.com/JuliaLang/julia/issues/43452#issuecomment-996971256>,
I found that currently the inlinear sometimes fails to handle `UnionAll`
call (e.g. runtime dispatch detected: Core.UnionAll(%28::TypeVar, %29::Any)).

This commit adds a missing late special handling for `UnionAll` calls:

> before
```julia
julia> code_typed((TypeVar,)) do tv
           UnionAll(tv, Type{tv})
       end
1-element Vector{Any}:
 CodeInfo(
1 ─ %1 = Core.apply_type(Main.Type, tv)::Type{Type{_A}} where _A
│   %2 = Main.UnionAll(tv, %1)::Any
└──      return %2
) => Any
```

> after
```julia
julia> code_typed((TypeVar,)) do tv
           UnionAll(tv, Type{tv})
       end
1-element Vector{Any}:
 CodeInfo(
1 ─ %1 = Core.apply_type(Main.Type, tv)::Type{Type{_A}} where _A
│   %2 = $(Expr(:foreigncall, :(:jl_type_unionall), Any, svec(Any, Any), 0, :(:ccall), Core.Argument(2), :(%1)))::Any
└──      return %2
) => Any
```

oscardssmith committed:
fix overflow bug in prevpow (#43410)

stevengj committed:
reverse iteration for eachline (#42225)

DilumAluthgeBot committed:
🤖 Bump the Pkg stdlib from 247a4062 to 43626542 (#43493)

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

SamuraiAku committed:
Add an SPDX file to the repository to streamline license and security reviews by user organizations (#42102)

* Add an SPDX file to the repository. 
* New script contrib/updateSPDX.jl .  Ran the script to update the SPDX file.

dkarrasch committed:
Merge pull request #43127 from JuliaLang/dk/freelinalg

Make LinearAlgebra.jl independent of SparseArrays.jl

vchuravy committed:
[LLVM] Drop support for LLVM 11 (#43371)

LilithHafner committed:
doc: we now have task migration (#43496)

Keno committed:
Make WINE overrideable for easier debugging (#43489)

* Make WINE overrideable for easier debugging

* Also use overriden wine for winepath

Otherwise things will ping-pong back and forth with the system
wine and never actually amke any progress.

Sean1708 committed:
Use correct delimiter for inline code. (#43524)

Originally ' had been used to close an inline code section, but it
should be `.

tkf committed:
RFC: Low-level API for `schedule` and `wait` (#41270)

aviatesk committed:
optimizer: improve SROA performance (#43488)

Essentially this commit reduces the runtime count # of domtree construction.
The key observation is that SROA of mutable allocations isn't often successful.
More specifically, during the sysimg build, `sroa_mutables!` is called
`5776` times, and actual SROA happened only for `126` cases out of those cases.
(, which means we should be able to avoid domtree construction in `5650`/`5776` cases).

aviatesk committed:
some compiler-related improvements and fixes (#43502)

* improve comipler type stabilities
* fix `code_typed_opaque_closure`

jakobnissen committed:
Improve inference for fallback TerminalMenu config access (#43533)

aviatesk committed:
add some NEWS entries for compiler-related changes (#43462)

timholy committed:
Add precompiles & disable constprop (#43367)

* Expand precompiles

These are targeted at reducing Revise latency if callbacks run
synchronously.

* Disable constprop on functions that don't benefit from it

This shaves about 100ms off the time to load Images.jl
under Revise.

tkf committed:
Fix a markdown link syntax error (#43543)

dkarrasch committed:
Remove obsolete method

dkarrasch committed:
Remove obsolete method

dkarrasch commented:
This requires a rebase, then this should work and we can merge.

mcognetta committed:
Merge branch 'dense_tridiag_mul' of github.com:mcognetta/julia into dense_tridiag_mul

mcognetta commented:
Ok, that did not work.

I am going to close this PR and open a new one with just the tests.

mcognetta commented:
Opened #43549 to resolve this.
